pkg/wire/util/size.go

```diff
@@ -0,0 +1,76 @@
+package util
+
+import (
+	"fmt"
+	"reflect"
+)
+
+// GetVarIntSize returns the size in number of bytes of a variable integer
+// (reference: GetVarSize(int value),  https://github.com/neo-project/neo/blob/master/neo/IO/Helper.cs)
+func GetVarIntSize(value int) int {
+	var size uintptr
+
+	if value < 0xFD {
+		size = 1 // unit8
+	} else if value <= 0xFFFF {
+		size = 3 // byte + uint16
+	} else {
+		size = 5 // byte + uint32
+	}
+	return int(size)
+}
+
+// GetVarStringSize returns the size of a variable string
+// (reference: GetVarSize(this string value),  https://github.com/neo-project/neo/blob/master/neo/IO/Helper.cs)
+func GetVarStringSize(value string) int {
+	valueSize := len([]byte(value))
+	return GetVarIntSize(valueSize) + valueSize
+}
+
+// GetVarSize return the size om bytes of a variable. This implementation is not exactly like the C#
+// (reference: GetVarSize<T>(this T[] value), https://github.com/neo-project/neo/blob/master/neo/IO/Helper.cs#L53) as in the C# variable
+// like Uint160, Uint256 are not supported. @TODO: make sure to have full unit tests coverage.
+func GetVarSize(value interface{}) int {
+	v := reflect.ValueOf(value)
+	switch v.Kind() {
+	case reflect.String:
```

Comment:
Reflection and Panics should be a last resort. For getting the size of a tx, you can call Bytes() on it and get the length()?

By the way, the `wire` package does not have many responsibilities; fee calculation would not be done at this level (Not sure if you were going to do it here, but just being clear), it is only responsibility for Decoding/Encoding and sending messages; this allows any application to import the `Wire` and communicate with the neo-network

Reply:
Unfortunately, I don't think it is that simple. The size calculation in the C# implementation is very much customized and it is used a basis for calculating the network fees. I don't think we can deviate here. Beside that the size of a transaction is also shown in the RPC calls and therefore users would expect to see the same size number between different implementation.

Reply:
Could we get some test vectors for this with transactions to see if they do differ and by how much?

To me, It seems that the main purpose of this functionality is to accurately represent how many bytes are being displayed in a transaction, which means they should match. The C# code seems to be the same, except for that the calculation is abstracted away in the Go code.

For example; a `ushort` is 16bits or 2 bytes, so  we see`element_size = 2` if the type is a ushort. I believe the additional values being added are from the integer encoding; `GetVarSize(value.Length) + value_size` 

This is all done under the cover when we tell Go what the type is and use binary.Write, which is why I believe they should match. Let me know if you find anything that differs.

Reply:
ok, will do. :+1: